deepreg/model/loss/image.py

```diff
@@ -50,167 +125,157 @@ def local_normalized_cross_correlation(
+        # the following is actually squared ncc
+        # shape = (batch, dim1, dim2, dim3, 1)
+        cross = tp_sum - p_avg * t_sum
+        t_var = t2_sum - t_avg * t_sum  # std[t] ** 2
+        p_var = p2_sum - p_avg * p_sum  # std[p] ** 2
+        ncc = (cross * cross + EPS) / (t_var * p_var + EPS)
+        return tf.reduce_mean(ncc, axis=[1, 2, 3, 4])
+
+    def get_config(self):
+        config = super(LocalNormalizedCrossCorrelation3D, self).get_config()
+        config["kernel_size"] = self.kernel_size
+        config["kernel_type"] = self.kernel_type
+        return config
+
+    @staticmethod
+    def build_kernel(kernel_size: int, kernel_type: str, ch: int):
+        """
+        :param kernel_size:
+        :param kernel_type:
+        :param ch:
+        :return: filters, kernel_vol
+        """
+        if kernel_type == "rectangular":
+            filters = tf.ones(shape=[kernel_size, kernel_size, kernel_size, ch, 1])
+            kernel_vol = kernel_size ** 3
+            return filters, kernel_vol
+
+        elif kernel_type == "triangular":
+            fsize = int((kernel_size + 1) / 2)
+            pad_filter = tf.constant(
+                [
+                    [0, 0],
+                    [int((fsize - 1) / 2), int((fsize - 1) / 2)],
+                    [int((fsize - 1) / 2), int((fsize - 1) / 2)],
+                    [int((fsize - 1) / 2), int((fsize - 1) / 2)],
+                    [0, 0],
+                ]
+            )
+
+            f1 = tf.ones(shape=(1, fsize, fsize, fsize, 1)) / fsize
+            f1 = tf.pad(f1, pad_filter, "CONSTANT")
+            f2 = tf.ones(shape=(fsize, fsize, fsize, 1, ch)) / fsize
+
+            filters = tf.nn.conv3d(f1, f2, strides=[1, 1, 1, 1, 1], padding="SAME")
+            filters = tf.transpose(filters, perm=[1, 2, 3, 4, 0])
+            kernel_vol = tf.reduce_sum(filters ** 2)
+
+            return filters, kernel_vol
+
+        elif kernel_type == "gaussian":
+            mean = (kernel_size - 1) / 2.0
+            sigma = kernel_size / 3
+
+            grid_dim = tf.range(0, kernel_size)
+            grid_dim_ch = tf.range(0, ch)
+            grid = tf.expand_dims(
+                tf.cast(
+                    tf.stack(tf.meshgrid(grid_dim, grid_dim, grid_dim, grid_dim_ch), 0),
+                    dtype="float32",
+                ),
+                axis=-1,
+            )
+            filters = tf.exp(
+                -tf.reduce_sum(tf.square(grid - mean), axis=0) / (2 * sigma ** 2)
+            )
+            kernel_vol = tf.reduce_sum(filters ** 2)
+
+            return filters, kernel_vol
+
+        else:
+            raise ValueError(
+                f"Wrong kernel_type for LNCC loss type. "
+                f"Please, specify a valid type 'rectangular' / 'triangular' / 'gaussian',"
+                f"got {kernel_type}"
+            )
+
+
+@REGISTRY.register_loss(name="lncc")
+class LocalNormalizedCrossCorrelation3DLoss(
```

Comment:
class names are not consistent: some with "3D", some without (energy/image/label all can be dimension-sensitive), while this is with "3DLoss"

Reply:
Those ones without 3D because they do not assume tensors have to be 3D.

Reply:
gradient and bending do assume 3d...

Reply:
oh there are some inconsistencies. I've fixed them :)